[jdbc] Perform upsert on conflict like the default for MySQL (PostgreSQL >=9.5)

henfiber opened the pull request:
This patch adds "upsert" behavior for PostgreSql >= 9.5, overwriting the previously stored value instead of raising an exception, when an existing timestamp is provided in a PUT request to the persistence endpoint.

This is the default behavior in the parent JdbcBaseDAO (and thus MySQL since it does not override it) and is equivalent to MySQL: `ON DUPLICATE KEY UPDATE VALUE`

see: https://www.postgresql.org/docs/9.5/sql-insert.html

This is only applied if the PostgreSQL version is >= 9.5, therefore it should not change the behavior for installations with legacy versions and should be backwards compatible with those versions.

Since `ModifiablePersistenceService` was only recently implemented, I would not consider it a breaking change either, rather than implementing consistent behavior between two major backends.

jlaur reviewed (approved):
Thanks for the fix!

henfiber commented:
Thanks for reviewing and merging. 

I will probably work more on the jdbc peristence addon, there are some inconsistencies in the handling of timestamps and time zones (I see also related open issues).

jlaur commented:
> I will probably work more on the jdbc peristence addon, there are some inconsistencies in the handling of timestamps and time zones (I see also related open issues).

That would be a very welcome addition. I recently did some work on JDBC persistence, but have been taking a break from it since the 3.4 release. 🙂 Was looking at #13775 most recently, you might want to check out the discussions there. AFAIR we didn't quite reach a conclusion.

henfiber commented:
I am not very experienced with MySQL so I am not sure I can help with this complex issue you have with the session and driver timestamps. 

I added a fix to the PostgreSQL service for now and added a PR here: https://github.com/openhab/openhab-addons/pull/14294 which also adds PG support to the console functionality you introduced at https://github.com/openhab/openhab-addons/pull/13765
